File: module2/class1/exemplo9.cs
/*
================================================================================
ATIVIDADE PRÁTICA 9 - BOXING/UNBOXING PERFORMANCE (C#)
================================================================================

OBJETIVO:
- Demonstrar overhead de boxing/unboxing em C#
- Usar Memory profiler para identificar alocações desnecessárias
- Otimizar usando generics e evitando boxing
- Medir impacto de boxing no GC e performance

PROBLEMA:
- Boxing cria objetos no heap para value types
- Unboxing requer cast e validação de tipo
- Memory Profiler mostrará alocações excessivas e GC pressure

SOLUÇÃO:
- Usar generics para evitar boxing
- Trabalhar com value types diretamente

================================================================================
*/

using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;

class Program {
    static void DemonstrateBoxingOverhead() {
        Console.WriteLine("Starting boxing/unboxing overhead demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see excessive allocations");
        
        const int ITERATIONS = 1000000;
        ArrayList boxingList = new ArrayList(); // Non-generic collection causes boxing
        
        var sw = Stopwatch.StartNew();
        long initialMemory = GC.GetTotalMemory(false);
        int initialCollections = GC.CollectionCount(0);
        
        // PERFORMANCE ISSUE: Boxing value types in non-generic collections
        for (int i = 0; i < ITERATIONS; i++) {
            boxingList.Add(i); // Boxing: int -> object allocation on heap
            
            if (i % 100000 == 0) {
                Console.WriteLine($"Boxed {i}/{ITERATIONS} integers...");
            }
        }
        
        // Now unbox them all
        long totalSum = 0;
        for (int i = 0; i < boxingList.Count; i++) {
            totalSum += (int)boxingList[i]; // Unboxing: object -> int with type check
        }
        
        sw.Stop();
        long finalMemory = GC.GetTotalMemory(false);
        int finalCollections = GC.CollectionCount(0);
        
        Console.WriteLine($"Boxing/unboxing completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total sum: {totalSum}");
        Console.WriteLine($"Memory used: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"GC collections: {finalCollections - initialCollections}");
        Console.WriteLine($"Boxed objects created: {ITERATIONS}");
    }
    
    static void Main() {
        Console.WriteLine("Starting boxing/unboxing performance demonstration...");
        Console.WriteLine("Task: Storing integers in non-generic collection");
        Console.WriteLine("Monitor Memory Usage Tool for boxing allocations");
        Console.WriteLine();
        
        DemonstrateBoxingOverhead();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check Memory profiler for:");
        Console.WriteLine("- High number of object allocations");
        Console.WriteLine("- Frequent GC collections");
        Console.WriteLine("- Memory pressure from boxing");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO SEM BOXING)
================================================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;

class Program {
    static void DemonstrateGenericCollection() {
        Console.WriteLine("Starting generic collection demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see reduced allocations");
        
        const int ITERATIONS = 1000000;
        List<int> genericList = new List<int>(); // CORREÇÃO: Generic collection avoids boxing
        
        var sw = Stopwatch.StartNew();
        long initialMemory = GC.GetTotalMemory(false);
        int initialCollections = GC.CollectionCount(0);
        
        // CORREÇÃO: No boxing - value types stored directly
        for (int i = 0; i < ITERATIONS; i++) {
            genericList.Add(i); // No boxing: int stays as value type
            
            if (i % 100000 == 0) {
                Console.WriteLine($"Added {i}/{ITERATIONS} integers (no boxing)...");
            }
        }
        
        // Direct access without unboxing
        long totalSum = 0;
        for (int i = 0; i < genericList.Count; i++) {
            totalSum += genericList[i]; // No unboxing: direct value access
        }
        
        sw.Stop();
        long finalMemory = GC.GetTotalMemory(false);
        int finalCollections = GC.CollectionCount(0);
        
        Console.WriteLine($"Generic collection completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total sum: {totalSum}");
        Console.WriteLine($"Memory used: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"GC collections: {finalCollections - initialCollections}");
        Console.WriteLine($"No boxing/unboxing occurred");
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized generic collection demonstration...");
        Console.WriteLine("Task: Storing integers in generic collection");
        Console.WriteLine("Monitor Memory Usage Tool for reduced allocations");
        Console.WriteLine();
        
        DemonstrateGenericCollection();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- No boxing allocations");
        Console.WriteLine("- Dramatically reduced memory usage");
        Console.WriteLine("- Fewer GC collections");
        Console.WriteLine("- Better performance due to value type semantics");
    }
}

================================================================================
*/

File: module2/class1/exemplo9.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 9 - OVERHEAD DE VIRTUAL FUNCTIONS (C++)
================================================================================

OBJETIVO:
- Demonstrar overhead de virtual function calls em loops intensivos
- Usar CPU profiler para identificar custos de virtual dispatch
- Otimizar usando templates ou função direta
- Comparar performance de virtual vs non-virtual calls

PROBLEMA:
- Virtual function calls requerem vtable lookup
- Indirect jumps impedem otimizações do compilador
- CPU Profiler mostrará tempo gasto em function call overhead

SOLUÇÃO:
- Usar templates para static dispatch
- Considerar direct function calls quando polimorfismo não é necessário

================================================================================
*/

#include <iostream>
#include <chrono>
#include <vector>
#include <memory>
using namespace std;

class VirtualProcessor {
public:
    virtual ~VirtualProcessor() = default;
    
    // PERFORMANCE ISSUE: Virtual function called in tight loops
    virtual int processValue(int value) = 0;
};

class ConcreteProcessor : public VirtualProcessor {
public:
    // Virtual function with vtable lookup overhead
    int processValue(int value) override {
        return value * value + 1; // Simple operation with virtual call overhead
    }
};

class DirectProcessor {
public:
    // Non-virtual function - direct call, compiler can inline
    int processValue(int value) {
        return value * value + 1; // Same operation without virtual overhead
    }
};

void demonstrateVirtualOverhead() {
    cout << "Starting virtual function overhead demonstration..." << endl;
    cout << "Monitor CPU profiler - should see virtual dispatch overhead" << endl;
    
    const int ITERATIONS = 10000000;
    vector<unique_ptr<VirtualProcessor>> processors;
    
    // Create processors
    for (int i = 0; i < 10; i++) {
        processors.push_back(make_unique<ConcreteProcessor>());
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long totalResult = 0;
    for (int i = 0; i < ITERATIONS; i++) {
        // PERFORMANCE BOTTLENECK: Virtual function call in tight loop
        int result = processors[i % processors.size()]->processValue(i);
        totalResult += result;
        
        if (i % 1000000 == 0) {
            cout << "Processed " << i << "/" << ITERATIONS << " virtual calls..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Virtual function processing completed in: " << duration.count() << " ms" << endl;
    cout << "Total result: " << totalResult << endl;
    cout << "Virtual calls made: " << ITERATIONS << endl;
}

int main() {
    cout << "Starting virtual function performance demonstration..." << endl;
    cout << "Task: Processing values using virtual function calls" << endl;
    cout << "Monitor CPU Usage Tool for virtual dispatch overhead" << endl << endl;
    
    demonstrateVirtualOverhead();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check CPU profiler for:" << endl;
    cout << "- Time spent in virtual function dispatch" << endl;
    cout << "- vtable lookup overhead" << endl;
    cout << "- Reduced compiler optimization opportunities" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO OTIMIZADA)
================================================================================

#include <iostream>
#include <chrono>
#include <vector>
using namespace std;

class DirectProcessor {
public:
    // CORREÇÃO: Non-virtual function allows compiler optimization
    inline int processValue(int value) {
        return value * value + 1; // Can be inlined by compiler
    }
};

template<typename ProcessorType>
void demonstrateDirectCalls() {
    cout << "Starting direct function call demonstration..." << endl;
    cout << "Monitor CPU profiler - should see reduced call overhead" << endl;
    
    const int ITERATIONS = 10000000;
    vector<ProcessorType> processors(10);
    
    auto start = chrono::high_resolution_clock::now();
    
    long long totalResult = 0;
    for (int i = 0; i < ITERATIONS; i++) {
        // CORREÇÃO: Direct function call - can be inlined, no vtable lookup
        int result = processors[i % processors.size()].processValue(i);
        totalResult += result;
        
        if (i % 1000000 == 0) {
            cout << "Processed " << i << "/" << ITERATIONS << " direct calls..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Direct function processing completed in: " << duration.count() << " ms" << endl;
    cout << "Total result: " << totalResult << endl;
    cout << "Function calls made: " << ITERATIONS << " (optimized)" << endl;
}

int main() {
    cout << "Starting optimized function call demonstration..." << endl;
    cout << "Task: Processing values using direct/template function calls" << endl;
    cout << "Monitor CPU Usage Tool for improved performance" << endl << endl;
    
    demonstrateDirectCalls<DirectProcessor>();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- No virtual function dispatch overhead" << endl;
    cout << "- Function inlining opportunities" << endl;
    cout << "- Better compiler optimizations" << endl;
    cout << "- Predictable branch patterns" << endl;
    
    return 0;
}

================================================================================
*/
